src/persist-client/src/usage.rs

```diff
@@ -29,10 +30,10 @@ impl StorageUsageClient {
     /// Creates a new StorageUsageClient pointed to a specific Blob
     pub async fn open(
         blob_uri: String,
-        client_cache: &mut PersistClientCache,
+        client_cache: Arc<Mutex<PersistClientCache>>,
```

Comment:
@pH14 - I'm definitely not confident that this is the right handling for this client, but I updated the storage usage type to look more like how I saw other clients interacting with the flavor of `PersistClientCache` that environmentd has access to. I'd love your feedback!

Reply:
ah this is a tricky bit for sure! the call to `open` requires a mutable reference to an existing `PersistClientCache`, and environmentd has an existing `Arc<Mutex<_>>` wrapper around our cache which lets us get that mutable reference -- the `Arc` makes it possible to share the reference among many owners (and in a thread-safe manner), and the `Mutex` means any one of those owner can get exclusive mutable access to the inner value, provided they're able to claim the lock.

what we can do is hoist the mutex locking up to the caller of this method instead, so `StorageUsageClient::persist` can continue to only ask for a mutable ref to `PersistClientCache`, and the caller can figure out how they want to provide a mutable ref.

in this case, we can actually leave `open` untouched, and then modify the caller to do something like:

```rust
    let storage_usage_response = {
        let mut client_cache = config.controller.persist_clients.lock().await;
        StorageUsageClient::open(
            config.controller.persist_location.blob_uri.clone(),
            &mut client_cache,
        )
        .await
    };
```

note that the lock will be held for the current scope of the call, or until an explicit `drop(client_cache)`. we can wrap the locking + creation of the client into its own `{` `}` scope so the lock is released right after creating the client

Reply:
Extremely helpful breakdown, thank you :+1: